cobc/tree.h

```diff
@@ -1792,6 +1796,12 @@ struct cb_ml_generate_tree {
 #define CB_ML_TREE(x)		(CB_TREE_CAST (CB_TAG_ML_TREE, struct cb_ml_generate_tree, x))
 #define CB_ML_TREE_P(x)	(CB_TREE_TAG (x) == CB_TAG_ML_TREE)
 
+struct cb_procedure_list {
+	struct cb_procedure_list	*next;
+	struct cb_procedure_list	*last;
```

Comment:
`last` is used in `codegen.c` to append new procedures at the end of the list in O(1).

Reply:
Ah, I now see `procedure_list_add()` uses that because the list is added "in order".

In _nearly_ all other places we use a stack: just add all entries (in the end the "current" one is the last one and has the previous one in `*prev` or `*next`) and in _many_ cases (the ones that do need it in order) there's a "reverse" block/function used when it is finished.

This makes setting the variables either (var->next = current, current = var). Could this apply here, too?
If you feel the "extra" entry for last and inserting it "in order" is more useful / needed, then I'm ok with that, too.

Reply:
I removed the `last` field of the list, and replaced it by a field `procedure_list_last` that always points to the last item. This way, `procedure_list` is in the good order to print, and new additions are done directly at its end.